configlib/config.go

```diff
@@ -38,9 +38,61 @@ func NewConfig(cfg *PkgrConfig) {
 		rVersion := rcmd.GetRVersion(&rs)
 		cfg.Library = getLibraryPath(cfg.Lockfile.Type, cfg.RPath, rVersion, rs.Platform, cfg.Library)
 	}
+
+	// For all cfg values that can be paths, make sure that ~ is expanded to the home directory.
+	cfg.Library = expandTilde(cfg.Library)
+	cfg.RPath = expandTilde(cfg.RPath)
+	cfg.Tarballs = expandTildes(cfg.Tarballs)
+	cfg.Repos = expandTildesRepos(cfg.Repos)
+	cfg.Logging.All = expandTilde(cfg.Logging.All)
+	cfg.Logging.Install = expandTilde(cfg.Logging.Install)
+	cfg.Cache = expandTilde(cfg.Cache)
+
 	return
 }
 
+/// expand the ~ at the beginning of a path to the home directory.
+/// consider any problems a fatal error.
+func expandTilde(p string) string {
+	expanded, err := homedir.Expand(p)
+	if err != nil {
+		log.WithFields(log.Fields{
+			"path": p,
+			"error": err,
+		}).Fatal("problem parsing config file -- could not expand path")
+	}
+	return expanded
+}
+
+/// For a list of paths, expand the ~ at the beginning of each path to the home directory.
+/// consider any problems a fatal error.
+func expandTildes(paths []string) []string {
+	var expanded []string
+	for _, p := range paths {
+		newPath := expandTilde(p)
```

Comment:
tldr; Package not needed. Personal opinion, use go's user lookup tools

This package isn't required anymore. In around go 1.9, 'nix systems will use /etc/passwd and more to locate the user details:

[Golang user docs](https://golang.org/pkg/os/user/)

> For most Unix systems, this package has two internal implementations of resolving user and group ids to names. One is written in pure Go and parses /etc/passwd and /etc/group. The other is cgo-based and relies on the standard C library (libc) routines such as getpwuid_r and getgrnam_r.

While it's fine to use for simplicity (IE you've already used) it may make more sense to implement the user looking and collection of homedir in-application. Especially on a project with such a low code velocity as the go-homedir one.

Reply:
I see so basically now we can just use os.Current()

which returns a pointer to a:

```
type User struct {
    // Uid is the user ID.
    // On POSIX systems, this is a decimal number representing the uid.
    // On Windows, this is a security identifier (SID) in a string format.
    // On Plan 9, this is the contents of /dev/user.
    Uid string
    // Gid is the primary group ID.
    // On POSIX systems, this is a decimal number representing the gid.
    // On Windows, this is a SID in a string format.
    // On Plan 9, this is the contents of /dev/user.
    Gid string
    // Username is the login name.
    Username string
    // Name is the user's real or display name.
    // It might be blank.
    // On POSIX systems, this is the first (or only) entry in the GECOS field
    // list.
    // On Windows, this is the user's display name.
    // On Plan 9, this is the contents of /dev/user.
    Name string
    // HomeDir is the path to the user's home directory (if they have one).
    HomeDir string
}
```

and just ask for `user.HomeDir`

That said, FWIW, I'd argue the low velocity makes go-homedir an easy choice since its well tested, a fixture of use, and keeps us from having to maintain (an admittedly small surface area) additional code. If anything, I'd fork it, change the implementation if we felt the need, then started using our version, rather than dropping it in each application.

Reply:
The other nice thing is that homedir automatically takes care of the actual string interpolation, which we'd have to do ourselves otherwise. Not particularly complicated, but it's something.